src/linkExposure.js

```diff
@@ -21,246 +25,458 @@ permissions.check({
+ * content script is running is in a private window.
+ * @param {string[]} linkExposureUpdate.linkUrls - The links on the page that the
+ * user was exposed to.
+ */
+function messageListener({ pageId, url, privateWindow, linkUrls }) {
+    // Increment the count of pending link exposure updates for the page
+    let pendingLinkExposureCount = pendingPageLinkExposureUpdates.get(pageId);
+    pendingLinkExposureCount = pendingLinkExposureCount === undefined ? 1 : pendingLinkExposureCount + 1;
+    pendingPageLinkExposureUpdates.set(pageId, pendingLinkExposureCount);
+
+    // Resolve all the link URLs in the update, converting each URL into a
+    // Promise<string>
+    const resolvedLinkUrlPromises = linkUrls.map((linkUrl) => {
+        return linkResolution.resolveUrl(linkUrl);
+    });
+
+    // Once resolution is complete, notify the linkExposureUpdate listeners
+    Promise.allSettled(resolvedLinkUrlPromises).then(async (results) => {
+        // For each link URL, if we have a resolved URL, use that
+        // If we don't have a resolved URL, use the original URL with
+        // cache, shim, and link decoration parsing
+        for(let i = 0; i < linkUrls.length; i++) {
+            if(results[i].status === "fulfilled") {
+                linkUrls[i] = results[i].value;
             }
             else {
-                await createLinkExposureRecord(linkExposure, nextLinkExposureIdCounter);
+                linkUrls[i] = await linkResolution.resolveUrl(linkUrls[i], { request: "none" });
             }
-        });
+        }
 
-    }, {
-        type: "webScience.linkExposure.exposureData",
-        schema: {
-            pageId: "string",
-            pageUrl: "string",
-            pageReferrer: "string",
-            pageVisitStartTime: "number",
-            privateWindow: "boolean",
-            nonmatchingLinkExposures: "number",
-            linkExposures: "object"
+        // If we are ignoring self links, determine whether each link URL is a self link
+        // by comparing to the page URL's public suffix + 1
+        // These are links that do not appear to be self links in the content
+        // script, but resolve to self links
+        let selfLinks = null;
+        if(ignoreSelfLinks) {
+            const pagePS1 = linkResolution.urlToPS1(url);
+            selfLinks = linkUrls.map(linkUrl => { return pagePS1 === linkResolution.urlToPS1(linkUrl); })
         }
-    });
 
-    initialized = true;
+        // Notify the listeners
+        for(const [listener, listenerRecord] of linkExposureUpdateListeners) {
+            // Check private window and page match pattern requirements for the listener
+            if((!privateWindow || listenerRecord.privateWindows) &&
+            listenerRecord.pageMatchPatternSet.matches(url)) {
+                const matchingLinkUrls = [];
+                let nonmatchingLinkCount = 0;
+                for(let i = 0; i < linkUrls.length; i++) {
```

Comment:
Same thing as above w/ `for ... of` and `linkUrls.entries()` iterator, if you want.